torchaudio/compliance/kaldi.py

```diff
@@ -0,0 +1,209 @@
+        pad_right = reversed_waveform
+        if pad > 0:
+            # torch.nn.functional.pad returns [2,1,0,1,2] for 'reflect'
+            # but we want [2, 1, 0, 0, 1, 2]
+            pad_left = reversed_waveform[-pad:]
+            waveform = torch.cat((pad_left, waveform, pad_right), dim=0)
+        else:
+            # pad is negative so we want to trim the waveform at the front
+            waveform = torch.cat((waveform[-pad:], pad_right), dim=0)
+
+    sizes = (m, window_size)
+    return waveform.as_strided(sizes, strides)
+
+
+def _feature_window_function(window_type, window_size, blackman_coeff):
+    """ Returns a window function with the given type and size
+    """
+    if window_type == HANNING:
+        return torch.hann_window(window_size, periodic=False)
+    elif window_type == HAMMING:
+        return torch.hamming_window(window_size, periodic=False, alpha=0.54, beta=0.46)
+    elif window_type == POVEY:
+        # like hanning but goes to zero at edges
+        return torch.hann_window(window_size, periodic=False).pow(0.85)
+    elif window_type == RECTANGULAR:
+        return torch.ones(window_size, dtype=torch.get_default_dtype())
+    elif window_type == BLACKMAN:
+        a = 2 * math.pi / (window_size - 1)
+        window_function = torch.arange(window_size, dtype=torch.get_default_dtype())
+        # can't use torch.blackman_window as they use different coefficients
+        return blackman_coeff - 0.5 * torch.cos(a * window_function) + \
+            (0.5 - blackman_coeff) * torch.cos(2 * a * window_function)
+    else:
+        raise Exception('Invalid window type ' + window_type)
+
+
+def _get_log_energy(strided_input, epsilon, energy_floor):
+    """ Returns the log energy of size (m) for a strided_input (m,*)
+    """
+    log_energy = torch.max(strided_input.pow(2).sum(1), epsilon).log()  # size (m)
+    if energy_floor == 0.0:
+        return log_energy
+    else:
+        return torch.max(log_energy,
+                         torch.tensor(math.log(energy_floor), dtype=torch.get_default_dtype()))
+
+
+def spectrogram(
+        sig, blackman_coeff=0.42, channel=-1, dither=1.0, energy_floor=0.0,
+        frame_length=25.0, frame_shift=10.0, min_duration=0.0,
+        preemphasis_coefficient=0.97, raw_energy=True, remove_dc_offset=True,
+        round_to_power_of_two=True, sample_frequency=16000.0, snip_edges=True,
+        subtract_mean=False, window_type=POVEY):
+    """Create a spectrogram from a raw audio signal. This matches the input/output of Kaldi's
+    compute-spectrogram-feats.
+
+    Inputs:
+        sig (Tensor): Tensor of audio of size (c, n) where c is in the range [0,2)
+        blackman_coeff (float): Constant coefficient for generalized Blackman window. (default = 0.42)
+        channel (int): Channel to extract (-1 -> expect mono, 0 -> left, 1 -> right) (default = -1)
+        dither (float): Dithering constant (0.0 means no dither). If you turn this off, you should set
```

Comment:
@cpuhrsch -- Do you know where this comment about `dither=0` and `energy_floor` comes from? This is the earliest I could track it to, and there is no discussion about it here.

Reply:
It's discussed in this function https://github.com/kaldi-asr/kaldi/blob/master/src/feat/feature-mfcc.h#L38. One of the suggested values is 1.0 for energy_floor. The default value for dither might have be chosen accidentally based on a misread, whereas for the tests it was set to 0 explicitly.

So we can choose a different default value without changing anything if it helps people

Reply:
The options come from here (https://github.com/kaldi-asr/kaldi/blob/master/src/feat/feature-window.h#L79). All the defaults in torchaudio match the defaults in Kaldi. If there is a flag that is not set in Kaldi and if its corresponding argument is not passed into the compliance function, they would have the same default value. It may seem strange if they were to mismatch as the user would not input all 16 flags into the compliance function or similarly, input all command line flags to the Kaldi executable.

Reply:
Thanks for the reply! It looks like kaldi does indeed use `dither=1.` as default, see [here](https://github.com/kaldi-asr/kaldi/blob/master/src/feat/feature-window.h#L57). We'll need to check if the results are different then.